### PR TITLE
Fixing the way the project update precentage is calculated

### PR DESCRIPTION
**Pull request by lancewf**
### :nut_and_bolt: Description: What code changed, and why?

There was a bug in the way we were calculating the percentage complete for a project update. We were just taking the last value from all the domain services. This change cleans up the percentage complete function to remove the cereal code to make the code easier to read. 
 
### :chains: Related Resources
#2436 

### :+1: Definition of Done

The percentage complete does not show 100% when the time is in the future. 

### :athletic_shoe: How to Build and Test the Change
1. `build components/authz-service && start_all_services`
1. Add a lot of data to compliance with `chef_load_compliance_scans -N1000 -D10 -M2 -T20000`
1. Create a project
1. Apply the rules
1. Ensure the percentage complete is not always 100%

### :white_check_mark: Checklist

- [x] Tests added/updated?
- [ ] Docs added/updated?

### CODE REVIEW

File: components/authz-service/server/v2/project_update_manager.go [hunk 1]
```diff
@@ -285,45 +284,20 @@ func (w *workflowInstance) PercentageComplete() float64 {
...
-		if subWorkflow.IsRunning() {
-			if err := subWorkflow.GetPayload(&payload); err != nil {
-				logrus.WithError(err).Errorf("failed to get payload for %q", d)
-				continue
-			}
-
-			if !payload.MergedJobStatus.Completed {
-				estimatedTime := time.Unix(payload.MergedJobStatus.EstimatedEndTimeInSec, 0)
-				if estimatedTime.After(longestEstimatedTimeComplete) {
-					percentComplete = float64(payload.MergedJobStatus.PercentageComplete)
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: Here is where the bug was. We should have been setting `longestEstimatedTimeComplete = estimatedTime`.

File: lib/authz/project_rules.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ func FindSlowestJobStatus(jobStatuses []JobStatus) JobStatus {
 	for _, jobStatus := range jobStatuses {
 		if !jobStatus.Completed {
 			combinedJobStatus.Completed = false
-			if jobStatus.EstimatedEndTimeInSec > combinedJobStatus.EstimatedEndTimeInSec {
+			if jobStatus.EstimatedEndTimeInSec >= combinedJobStatus.EstimatedEndTimeInSec {
```
Comment (lancewf) [inline lib/authz/project_rules.go]: This allows for the case when no statuses have a time estimate.

File: components/authz-service/server/v2/project_update_manager.go [hunk 3]
```diff
@@ -285,45 +284,28 @@ func (w *workflowInstance) PercentageComplete() float64 {
...
 
-	percentComplete := 0.0
-	domainServicesUpdateInstance, err := w.GetUpdateDomainServicesInstance()
-	if err == cereal.ErrWorkflowInstanceNotFound {
-		return percentComplete
-	}
-
-	domainServices := domainServicesUpdateInstance.ListSubWorkflows()
-	if len(domainServices) == 0 {
+	domainJobStatuses := w.collectAllDomainServiceJobStatuses()
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: Below is what the function looks like in this PR. 
```
func (w *workflowInstance) PercentageComplete() float64 {
	if !w.IsRunning() {
		return 1.0
	}

	domainJobStatuses := w.collectAllDomainServiceJobStatuses()
	if len(domainJobStatuses) == 0 {
		return 1.0
	}

	longestDomainJobStatus := project_update_tags.FindSlowestJobStatus(domainJobStatuses)

	return float64(longestDomainJobStatus.PercentageComplete)
}
```

File: components/authz-service/server/v2/project_update_manager_internal_test.go [hunk 4]
```diff
@@ -1,51 +0,0 @@
-package v2
-
-import (
-	"testing"
-	"time"
-
-	"github.com/stretchr/testify/assert"
-)
-
-func TestFindLongestEstimatedTimeEmpty(t *testing.T) {
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager_internal_test.go]: This function was removed for the already existing `FindSlowestJobStatus` function.

File: components/authz-service/server/v2/project_update_manager.go [hunk 5]
```diff
@@ -408,27 +390,14 @@ func (m *CerealProjectUpdateManager) Status() (ProjectUpdateStatus, error) {
 	return projectUpdateInstance, nil
 }
 
-// Find the longest estimate that is in the future. If no estimate is found return zero time.Time{}
-func findLongestEstimatedTime(estimatedTimes []time.Time) time.Time {
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: Replaced with the already existing `FindSlowestJobStatus` function.

File: components/authz-service/server/v2/project_update_manager.go [hunk 6]
```diff
@@ -408,27 +390,14 @@ func (m *CerealProjectUpdateManager) Status() (ProjectUpdateStatus, error) {
...
-			longestEstimatedTimeComplete = estimatedTime
-		}
-	}
-
-	return longestEstimatedTimeComplete
-}
-
-func (w *workflowInstance) collectAllDomainServiceEstimates() []time.Time {
-	estimates := make([]time.Time, 0)
+func (w *workflowInstance) collectAllDomainServiceJobStatuses() []project_update_tags.JobStatus {
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: All the cereal logic is hidden in this function now.

**Comment by msorens** [comment]
Some observations:
1. ❓Compliance percentage is **not** strictly monotonic increasing. Sample: 15%, 78%, 63%, 100%
2. ❓Compliance estimated time is **not** strictly monotonic increasing. Sample: 1576622615, 1576622536, 1576622570, 1576622537, 1576622565, 1576622537
3. ✔️ Compliance completed flag remains at `false` until done, then remains at `true`. 
4. ❓Compliance estimated completion time may still change after `Completed` becomes `true`.
5. ❌Compliance percentage can decrease after touching 100%. Sample: 15%, 100%, 93%, 100% (Note that `Completed` only became `true` on the **final** 100) 

Not sure if there is any significance but the logging calls show that `JobStatus`, whenever it is called, is invoked **twice**--and the reported values are different. If, for example, I only look at the first call of each pair, then several of the issues above go away.

**Comment by lancewf** [comment]
@msorens 
❓Compliance percentage is not strictly monotonic increasing. Sample: 15%, 78%, 63%, 100%
**True**
❓Compliance estimated time is not strictly monotonic increasing. Sample: 1576622615, 1576622536, 1576622570, 1576622537, 1576622565, 1576622537
**True**
✔️ Compliance completed flag remains at false until done, then remains at true.
**True**
❓Compliance estimated completion time may still change after Completed becomes true.
**false** _I have not seen this behavior._ 
❌Compliance percentage can decrease after touching 100%. Sample: 15%, 100%, 93%, 100% (Note that Completed only became true on the final 100)
**True**

The Completed flag is the surest value. Where the percentage complete and the estimated time fluctuate when the project update is first starting. The estimated time is calculated of the percentage complete. 

Note: This PR is only trying to fix the problem that we were only displaying the percentage complete of the last job and not the job that is predicted to run the longest.

**Comment by susanev** [comment]
@lancewf do you mind opening up a separate issue around 

> Note: This PR is only trying to fix the problem that we were only displaying the percentage complete of the last job and not the job that is predicted to run the longest.

**Comment by lancewf** [comment]
Added https://github.com/chef/automate/issues/2540

**Comment by msorens** [comment]
My observations were to try to determine if there are still other bugs requiring attention; I think you are saying that there are not. 
However, the one that concerns me is that the percentage might go to 100% even when `Completed` is still `false`. Seems like that directly ties in with the topic of this PR.

**Pull request merged**